pkg/health/condition/check_backup_ready.go

```diff
@@ -37,101 +38,194 @@ const (
+	// NotChecked is a constant that means that the etcd backup status has not been updated or rechecked
+	NotChecked string = "ConditionNotChecked"
 )
 
 func (a *backupReadyCheck) Check(ctx context.Context, etcd druidv1alpha1.Etcd) Result {
-	//Default case
-	result := &result{
-		conType: druidv1alpha1.ConditionTypeBackupReady,
-		status:  druidv1alpha1.ConditionUnknown,
-		reason:  Unknown,
-		message: "Cannot determine etcd backup status",
-	}
-
 	// Special case of etcd not being configured to take snapshots
 	// Do not add the BackupReady condition if backup is not configured
 	if etcd.Spec.Backup.Store == nil || etcd.Spec.Backup.Store.Provider == nil || len(*etcd.Spec.Backup.Store.Provider) == 0 {
 		return nil
 	}
 
-	//Fetch snapshot leases
-	var (
-		fullSnapErr, incrSnapErr error
-		fullSnapLease            = &coordinationv1.Lease{}
-		deltaSnapLease           = &coordinationv1.Lease{}
+	// Fetch snapshot leases
+	fullSnapshotLease, err := a.fetchLease(ctx, etcd.GetFullSnapshotLeaseName(), etcd.Namespace)
+	if err != nil {
+		return createBackupConditionResult(
+			druidv1alpha1.ConditionUnknown, Unknown,
+			fmt.Sprintf("Unable to fetch full snap lease. %s", err.Error()),
+		)
+	}
+
+	deltaSnapSnapshotLease, err := a.fetchLease(ctx, etcd.GetDeltaSnapshotLeaseName(), etcd.Namespace)
+	if err != nil {
+		return createBackupConditionResult(
+			druidv1alpha1.ConditionUnknown, Unknown,
+			fmt.Sprintf("Unable to fetch delta snap lease. %s", err.Error()),
+		)
+	}
+
+	deltaSnapshotLeaseRenewTime := deltaSnapSnapshotLease.Spec.RenewTime
+	fullSnapshotLeaseRenewTime := fullSnapshotLease.Spec.RenewTime
+	fullSnapshotLeaseCreationTime := &fullSnapshotLease.ObjectMeta.CreationTimestamp
+	fullSnapshotDuration, err := utils.ComputeScheduleDuration(*etcd.Spec.Backup.FullSnapshotSchedule)
+	if err != nil {
+		return createBackupConditionResult(
+			druidv1alpha1.ConditionUnknown, Unknown,
+			fmt.Sprintf("Unable to compute full snapshot duration from schedule. %v", err.Error()),
+		)
+	}
+
+	// Both snapshot leases are not yet renewed
+	if fullSnapshotLeaseRenewTime == nil && deltaSnapshotLeaseRenewTime == nil {
+		return createBackupConditionResult(
+			druidv1alpha1.ConditionUnknown, Unknown,
+			"Snapshotter has not started yet",
+		)
+	}
+
+	// Most probable during reconcile of existing clusters if fresh leases are created
+	if fullSnapshotLeaseRenewTime == nil && deltaSnapshotLeaseRenewTime != nil {
+		return handleOnlyDeltaSnapshotLeaseRenewedCase(
+			fullSnapshotLeaseCreationTime.Time, fullSnapshotDuration,
+			deltaSnapshotLeaseRenewTime.Time, 2*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration,
+		)
+	}
+
+	// Most probable during a startup of a new cluster and only full snapshot has been taken
+	if fullSnapshotLeaseRenewTime != nil && deltaSnapshotLeaseRenewTime == nil {
+		return handleSnapshotterStartupCase(
+			fullSnapshotLease.Spec.RenewTime.Time,
+			5*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration,
```

Comment:
Can we have create functions for delta and full snapshot grace period? These can be methods on etcd resource itself, if you do not like that then these can just be standalone functions. One single place to compute these preventing duplicating these multipliers in more than one place in future code changes.

Reply:
If you notice, the grace period isn't same across functions. Grace period for renewing delta snapshot lease that was already renewed previously is 2x delta snapshot period, while grace period for delta snapshotting to start once full snapshot has been taken is 5x delta snapshot period. So it's not achievable by one single function. Also, this is condition-specific information, and may not make sense to be part of the API, because the API has no knowledge of which conditions are set on the etcd status by druid, and it shouldn't as well.